test/algorithms/iter/bdl_input_iterator.cpp

```diff
@@ -0,0 +1,289 @@
+    bii += 2ull;
+
+    CHECK(bii == 2ull);
+    CHECK(bii != 1ull);
+
+    CHECK(bii < 3ull);
+    CHECK(bii <= 3ull);
+
+    CHECK(bii > 1ull);
+    CHECK(bii >= 1ull);
+
+    const auto bii_cp = bii;
+
+    // decrement assignment
+    bii -= 2ull;
+
+    CHECK(bii == 0ull);
+    CHECK(bii != 1ull);
+
+    CHECK(bii < 1ull);
+    CHECK(bii <= 1ull);
+
+    CHECK(bii >= 0ull);
+
+    // difference
+    CHECK(bii_cp - bii == 2);
+
+    // subscript
+    CHECK(bii[0] == 0ull);
+    CHECK(bii[1] == 1ull);
+    CHECK(bii[2] == 2ull);
+    CHECK(bii[3] == 3ull);
+    CHECK(bii[4] == 4ull);
+}
+
+TEST_CASE("Empty layout iteration", "[bdl-input-iterator]")
+{
+    using layout = sidb_cell_clk_lyt_siqad;
+
+    layout lyt{};
+
+    bdl_input_iterator<layout> bii{lyt};
+
+    CHECK((*bii).num_cells() == 0);
+
+    // increment
+
+    ++bii;
+
+    CHECK((*bii).num_cells() == 0);
+
+    auto bii_cp = bii++;
+
+    CHECK((*bii).num_cells() == 0);
+    CHECK((*bii_cp).num_cells() == 0);
+
+    // decrement
+
+    --bii;
+
+    CHECK((*bii).num_cells() == 0);
+
+    auto bii_cm = bii--;
+
+    CHECK((*bii).num_cells() == 0);
+
+    CHECK((*bii_cm).num_cells() == 0);
+}
+
+TEST_CASE("BDL wire iteration", "[bdl-input-iterator]")
+{
+    using layout = sidb_cell_clk_lyt_siqad;
+
+    layout lyt{{20, 0}, "BDL wire"};
+
+    lyt.assign_cell_type({0, 0, 0}, sidb_technology::cell_type::INPUT);
+    lyt.assign_cell_type({2, 0, 0}, sidb_technology::cell_type::INPUT);
+
+    lyt.assign_cell_type({6, 0, 0}, sidb_technology::cell_type::NORMAL);
+    lyt.assign_cell_type({8, 0, 0}, sidb_technology::cell_type::NORMAL);
+
+    lyt.assign_cell_type({12, 0, 0}, sidb_technology::cell_type::NORMAL);
+    lyt.assign_cell_type({14, 0, 0}, sidb_technology::cell_type::NORMAL);
+
+    lyt.assign_cell_type({18, 0, 0}, sidb_technology::cell_type::OUTPUT);
+    lyt.assign_cell_type({20, 0, 0}, sidb_technology::cell_type::OUTPUT);
+
+    bdl_input_iterator<layout> bii{lyt};
+    CHECK(bii == 0ull);
+
+    // start by incrementing over all input states
+
+    // layout at input state 0
+    const auto& lyt_0 = *bii;
+
+    // the iterator should have toggled the second input cell
+    CHECK(lyt_0.get_cell_type({0, 0, 0}) == sidb_technology::cell_type::INPUT);
+    CHECK(lyt_0.get_cell_type({2, 0, 0}) == sidb_technology::cell_type::EMPTY);
+
+    ++bii;
+    CHECK(bii == 1ull);
+
+    // layout at input state 1
+    const auto& lyt_1 = *bii;
+
+    // the iterator should have toggled the first input cell
+    CHECK(lyt_1.get_cell_type({0, 0, 0}) == sidb_technology::cell_type::EMPTY);
+    CHECK(lyt_1.get_cell_type({2, 0, 0}) == sidb_technology::cell_type::INPUT);
+
+    // doing another iteration should overflow and set it back to 0
+    ++bii;
+    CHECK(bii == 2ull);
+
+    const auto& lyt_2 = *bii;
+
+    CHECK(lyt_2.get_cell_type({0, 0, 0}) == sidb_technology::cell_type::INPUT);
+    CHECK(lyt_2.get_cell_type({2, 0, 0}) == sidb_technology::cell_type::EMPTY);
+
```

Comment:
What happens if I increment again? Is this something to test?

Reply:
We would get to input state 1 again since we have completely wrapped around. This very wrap-around test is already intended to check this particular behavior.

Reply:
True! Sorry.

Reply:
No worries :smile: